### ISSUE TITLE
Block Styles Breakdown

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#20331]
This is an overview of the concrete tasks needed to proceed with the project scope of https://github.com/WordPress/gutenberg/issues/9534 (see also https://github.com/WordPress/gutenberg/issues/19611). It operates on three levels, or style origins: local blocks, theme defaults, and global modifications.

<img width="547" alt="Screenshot 2020-03-11 at 19 59 19" src="https://user-images.githubusercontent.com/583546/83249402-caf63e80-a1a6-11ea-8b2e-7fa72473aa3d.png">

## First 🌊

<details>

<summary>Goal: prototype a system that connects the three style origins and demonstrates how it works for a few top-level blocks.</summary>

### Infrastructure

_Ensure at all times the editor reflects the front faithfully._

- [x] Create the concept of implicit style attributes to control the appearance of the block. #15450 #21021
- [x] Centralize style mappings https://github.com/WordPress/gutenberg/issues/25051
- [x] Migrate existing attributes (`font-size`, `color`) and introduce missing properties (`line-height`, `family`).  #22700
  - [x] font-size: paragraph (#21153), heading (#21431)
  - [x] line-height: paragraph (#20775), heading (#21215)
  - [x] color: paragraph (#21037), heading (#21039)
  - [x] set colors individually https://github.com/WordPress/gutenberg/issues/24927
  - [x] weight https://github.com/WordPress/gutenberg/pull/24978
  - [x] family https://github.com/WordPress/gutenberg/issues/23204
- [x] Evolve a `theme.json` specification for controlling the editor and the style attributes of different origins. #22518 #20588
- [x] Merge the different style origins (block, theme, user). #20047 #20290
- [x] Generate front-end styles using the new system. #19883
- [x] Generate editor styles using the new system. #22520

### Interface

_These are the different tools that interact with appearance values._

- [x] Expose global styles tool in `edit-site`. This should be a straightforward rendering of a sidebar in the site editor displaying the global values and palette. #24250 
- [x] Allow users to modify some style attributes in the global styles panel. #20367
  - [x] colors: link colors (link, hover / focus, active) #21032
  - [x] typography: font families, font size, line height. #21030 #21028 #20773
  - [x] reset modifications to theme defaults. #20868
- [x] Develop interface components to operate with new attributes. Prior art in font sizes and color panel. #20367
  - [x] line-height to Paragraph and Heading inspectors. #20339

</details>

## Second 🌊

<details>
<summary>Goal: develop all necessary sub-systems.</summary>

### Infrastructure

- [x] Stabilize:
  - [x] (one pr remaining to merge) Client: review there's no duplication and improve the way metadata is structured.
  - [x] Server-client metadata: do not send data we already have https://github.com/WordPress/gutenberg/issues/26802
  - [x] Server: consolidate theme.json entity https://github.com/WordPress/gutenberg/issues/26802
  - [x] Server: consolidate theme.json processor https://github.com/WordPress/gutenberg/pull/27237
  - [x] ~Prepare a core patch for 5.7 (closed)~ https://github.com/WordPress/gutenberg/issues/27506
- [x] theme.json
  - [x] Consolidate dynamic selectors (selectors for blocks that represent multiple HTML elements, such as heading or list). https://github.com/WordPress/gutenberg/pull/24423 https://github.com/WordPress/gutenberg/pull/26945
  - [x] ~Control design tools visibility (superseded)~ https://github.com/WordPress/gutenberg/issues/27295
  - [x] Make settings & styles top-level keys https://github.com/WordPress/gutenberg/issues/28163
  - [x] Split `global` selector into `defaults` and `root` https://github.com/WordPress/gutenberg/pull/28533
/24165.
- [x] Style properties
  - [x] font-style https://github.com/WordPress/gutenberg/pull/26444 https://github.com/WordPress/gutenberg/pull/26844
  - [x] font-weight https://github.com/WordPress/gutenberg/pull/26444 https://github.com/WordPress/gutenberg/pull/26844
  - [x] text-decoration https://github.com/WordPress/gutenberg/pull/26059
  - [x] text-transform https://github.com/WordPress/gutenberg/pull/26060
  - [x] Padding (global UI + theme.json) https://github.com/WordPress/gutenberg/pull/27099 https://github.com/WordPress/gutenberg/pull/27154
  - [x] border-radius https://github.com/WordPress/gutenberg/pull/25791
- [x] Presets
  - [x] Make them translatable via theme.json https://github.com/wp-cli/i18n-command/issues/224 https://github.com/WordPress/gutenberg/pull/28246
  - [x] Allow creating presets in contexts other than `global`. https://github.com/WordPress/gutenberg/pull/25768
  - [x] Allow creating user-defined color palettes. https://github.com/WordPress/gutenberg/pull/27250
  - [x] Allow presets to use other preset variables https://github.com/WordPress/gutenberg/issues/26511

### Interface

- [x] Make editor controls reflect styles, whether they come from theme or user. https://github.com/WordPress/gutenberg/pull/21490 https://github.com/WordPress/gutenberg/pull/22722 https://github.com/WordPress/gutenberg/pull/26783 https://github.com/WordPress/gutenberg/pull/26786
- [x] Reset to theme defaults https://github.com/WordPress/gutenberg/pull/26975 

</details>

## Third 🌊

<details>

<summary>Goal: merge the non UI parts of global styles in WordPress 5.8.</summary>

block.json / block supports:

- [x] https://github.com/WordPress/gutenberg/issues/28913 Add support for skip serialization to all properties. Core blocks can be migrated iteratively.
- [x] https://github.com/WordPress/gutenberg/issues/29616 `border-width`, `border-style`, `border-color`
- [x] https://github.com/WordPress/gutenberg/pull/29335 Alignments
- [x] https://github.com/WordPress/gutenberg/issues/31488 link color: make it work like layout or duotone. https://github.com/WordPress/gutenberg/issues/25151 Will also fix link color for author roles.

theme.json:

- [x] https://github.com/WordPress/gutenberg/issues/29891 Changes to theme.json
- [x] Versioning theme.json from the theme, implemented as part of https://github.com/WordPress/gutenberg/pull/30541
- [x] https://github.com/WordPress/gutenberg/issues/29568 colors: allow having defaults, theme, and custom
- [x] https://github.com/WordPress/gutenberg/issues/29778 Theme authors should be able to use any styles in theme.json independently from the block.json support. 
- [x] https://github.com/WordPress/gutenberg/issues/24165 Mobile: confirm there are not blockers to make theme.json cross-platform (web & mobile).
- [x] https://github.com/WordPress/gutenberg/issues/30191 Use `wp_theme` taxonomy instead of post_name to associate user styles with particular themes.

Internationalization:

- [x] https://github.com/WordPress/gutenberg/pull/29828 theme.json: customTemplates.
- [x] block supports: allow changing UI control labels https://github.com/WordPress/gutenberg/pull/29155 (discarded, instead we rephrased the labels to be more adequate for different contexts) https://github.com/WordPress/gutenberg/pull/30075

Docs:

- [x] Update [theme.json docs](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/) to make clear what's part of WordPress 5.8 and what's part of the plugin. https://github.com/WordPress/gutenberg/pull/33131
- [x] Update [theme support docs](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-support/) to state that link color is now stable and part of Wordpress core, using `theme.json`. The `experimental-link-color` theme support will be deprecated and removed when the plugin requires WordPress 5.8 as the minimum version. https://github.com/WordPress/gutenberg/pull/33162
- [x] Update docs for block supports, including a note about `__experimentalSkipSerialization`.

Extensibility:

- [x] https://github.com/WordPress/gutenberg/issues/27504 Hooking into the lifecycle of GS: stabilize API and define what sort of hooks do we need.
- [x] https://github.com/WordPress/gutenberg/issues/30082 Block editor settings & global styles settings.

Bug fixes

- [x] Disable/Enable custom colors, gradients, and font sizes https://github.com/WordPress/gutenberg/pull/32200
- [x] Empty color palette should disable the feature https://github.com/WordPress/gutenberg/pull/32225 https://github.com/WordPress/gutenberg/pull/32358
- [x] `settings.color` no longer have the theme colors https://github.com/WordPress/gutenberg/issues/32027 https://github.com/WordPress/gutenberg/pull/32358
- [x] Block presets generate incorrect styles https://github.com/WordPress/gutenberg/issues/31660

</details>

## Fourth 🌊

Follow-ups to WordPress 5.8 release.

Block supports:

- [x] https://github.com/WordPress/gutenberg/issues/31489

theme.json:

- [x] https://github.com/WordPress/gutenberg/issues/34414
- [x] https://github.com/WordPress/gutenberg/issues/34349
- [ ] https://github.com/WordPress/gutenberg/issues/33470
- [ ] https://github.com/WordPress/gutenberg/issues/33500
- [ ] https://github.com/WordPress/gutenberg/issues/34317
- [x] https://github.com/WordPress/gutenberg/issues/34507

Stylesheet:

- [x] https://github.com/WordPress/gutenberg/issues/34141

Performance:

- [x] https://github.com/WordPress/gutenberg/issues/34328
- [x] Only ship default presets once:
  - [x] https://github.com/WordPress/gutenberg/issues/34138
  - [x] https://github.com/WordPress/gutenberg/issues/34006
- [ ] https://github.com/WordPress/gutenberg/issues/31642
- [ ] https://github.com/WordPress/gutenberg/issues/33977
- [ ] https://github.com/WordPress/gutenberg/issues/34329

Interface

- [ ] https://github.com/WordPress/gutenberg/issues/34574

## Backlog

The things listed here don't need to be necessarily implemented, but should be considered.

block.json: nothing planned

- https://github.com/WordPress/gutenberg/issues/27614 Potential use case to add height (spacer and cover, which has a custom one at the moment).

theme.json

- https://github.com/WordPress/gutenberg/issues/33232 See also the [use case for margin](https://github.com/WordPress/gutenberg/pull/22208#issuecomment-644099875)
- Versioning theme.json data from the user https://github.com/WordPress/gutenberg/issues/27230#issuecomment-796884408 
- Consider a standardized way to modify hover/focus/active states for blocks https://github.com/WordPress/gutenberg/issues/27075
- Allow all block-editor theme support options to be defined via theme.json #26901
- Consider allowing themes to enqueue stylesheets in theme.json #26902
- How to express nested content
  - Use a flat structure of selectors `core/group core/paragraph`. https://github.com/WordPress/gutenberg/issues/28163
  - We could have separate files for template parts (in addition to the theme.json for the site or main "document", see https://github.com/WordPress/gutenberg/issues/27233
  - We could create CSS vars for container blocks https://github.com/WordPress/gutenberg/pull/29714#issuecomment-796591814
- Do we need to support dynamic values that change depending on other values? Example for colors https://github.com/WordPress/gutenberg/issues/24166
- Consolidate block attributes, block supports, and theme.json shape https://github.com/WordPress/gutenberg/issues/22887
- Schema and type validation https://github.com/WordPress/gutenberg/issues/26955
- Allow CSS properties that we don't manage? https://github.com/WordPress/gutenberg/issues/27231
- Responsive styles/media queries (styles depending on window size). https://github.com/WordPress/gutenberg/issues/20244
- Define if global attributes (base `font`, `color`, page `background`) can be conceptualized as attributes of a root Site block. #16998 Related: consider global site width and padding #20771

Expand to more style properties:

- text alignment https://github.com/WordPress/gutenberg/issues/24616
- `height` https://github.com/WordPress/gutenberg/pull/28409
- `margin` https://github.com/WordPress/gutenberg/pull/25988
- `width` https://github.com/WordPress/gutenberg/issues/31037
- Free-form custom CSS area. Consider whether this should be per-block as in https://github.com/WordPress/gutenberg/pull/25413 or global to all of them?
- API for loading font families. This should be inherently extensible and likely unopinionated. Maybe explore an API like `wp_register_font_family`.
- Justified text https://github.com/WordPress/gutenberg/issues/27300

Extensibility:

- Support for child themes https://github.com/WordPress/gutenberg/issues/25612
- Export `theme.json` configs (including user preferences) as part of exporting templates & template parts https://github.com/WordPress/gutenberg/issues/27941 https://github.com/WordPress/gutenberg/issues/27528
- UI extensibility: allow 3rd parties to hook into it, etc.

Interface:

- Global styles sidebar: labels for panels in the global styles sidebar (the subtitle of the "headings" panel, for example, "Headings (h1)", is taken from the `__experimentalSelector` of the heading block). How is this in the new UI?
- Integrate new component system https://github.com/WordPress/gutenberg/issues/28399 https://status.g2-components.com/
- Do not show-sidebar if the theme doesn't have theme.json support https://github.com/WordPress/gutenberg/issues/27394
- Custom controls for the Global Styles sidebar. https://github.com/WordPress/gutenberg/issues/29778
- Themes can control the visibility of UI controls. Relates to sidebars https://github.com/WordPress/gutenberg/issues/27331 and the old proposal to control visibility https://github.com/WordPress/gutenberg/issues/27295 If something is disabled via theme.json it should be disabled for both the global styles and the block sidebar.
- Blue dot on block panels that have modifications https://github.com/WordPress/gutenberg/issues/27474
- Flow for editing the specific styles of a specific block type https://github.com/WordPress/gutenberg/issues/27393
- Advanced panels of blocks should show which Global Styles are affecting it https://github.com/WordPress/gutenberg/issues/27475
- Exploring beyond a sidebar for exposing styling tools  #20212
- Explore combining all style-related attributes of a block into a "Style" tab in the inspector separate from "Settings".
- Explore a way to lift local modifications on blocks to global (apply certain modifications to all paragraphs or to all blocks that consume an attribute).
- Explore if it's needed to show locally on blocks when they are inheriting from a global or theme setting.
- Allow granular saving options https://github.com/WordPress/gutenberg/issues/29388

Iterate on design tools:

- Add background editing tools to Cover. #20193
- Add spacing control to Gallery. #20705
- Establish background tools components hooked to the relevant attributes. #16479
- Look at how to absorb units for font-size https://github.com/WordPress/gutenberg/issues/25137 https://github.com/WordPress/gutenberg/issues/23323

Misc:

- Try using transient attributes on global styles entity https://github.com/WordPress/gutenberg/issues/25138
- Dark mode: editor https://github.com/WordPress/gutenberg/issues/28200 authors https://github.com/WordPress/gutenberg/issues/24368
- Do not create revisions for the CPT used by global styles https://github.com/WordPress/gutenberg/issues/30132

**Comment by jorgefilipecosta** [comment]
Sharing some thoughts on the current status for infrastructure/ my interpretation of some infrastructure tasks:

> Create the concept of implicit attributes for appearance pairing 

It seems the style attribute we are using for colors introduced in https://github.com/WordPress/gutenberg/pull/21021 addressed this part we are also using it for lineHeight with success and without the need to register new attributes.

>  Hook appearance variables to the editor rendering. 

It seems this is about making sure styles are regenerated when the structure that holds global style changes. For example, https://github.com/WordPress/gutenberg/pull/20530/ accomplishes this by compiling the styles on GlobalStylesProvider.

>  Look at style variations as packages of style attribute sets.

We can achieve this by adding a styles structure to theme.json:
{
    ...
    styles: {
        ‘core/paragraph’: {
            ‘huge’: {
                fontSize: 14,
            }
        }
    }
}

We may also achieve the "Custom User Styles" part of https://github.com/WordPress/gutenberg/issues/7551 by just allowing the user to save the current block style attributes as a customization of the theme.json structure.

**Comment by kjellr** [comment]
At this week's Block-based Themes meeting, the group discussed some of the upcoming design tools in this ticket, and whether they should be opt-in or opt-out for themes. Here are our notes: 

- If a theme has `theme.json`, it seems ok to have all new design tools be opt-out. 
- Otherwise, if the feature is expected to “just work” with themes without breaking something, it’s probably ok to enable it by default. Otherwise, they should be opt-in.
- If these are “editor” features, and not something the theme needs to explicitly build in support for, then it may be confusing to use `add_theme_support()` for them. 
- We should keep in mind that if users move between block-based themes and non-block-based themes, it will likely be confusing for seemingly unrelated features to turn on and off.

More details in [the meeting notes](https://make.wordpress.org/themes/2020/10/07/block-based-meeting-notes-oct-7-2020/).

**Comment by jorgefilipecosta** [comment]
With the first beta version of WordPress 5.7 not very far away I'm sharing this summary after some discussion with @nosolosw specifying what are the tasks we think should be part of v1 and/or taht may be part of v1. 

## What we will do next?

These are the tasks we will work on and plan to have ready soon. Are the most prioritary tasks.

-   Versioning.
    

	-   This needs exploration. One option is that theme.json files have a version number, similarly to block.json files. so if we need to make a breaking change we can still keep code that is able to deal with files from a previous version.
    

-   Child theme support.
    

	-   A child theme inherits theme.json from parents(s) can add new properties or overwrite existing properties.
    

-   (In progress) Padding support on global styles.
    
-   (In progress) Refactor our code and merge it into the core.
    
-   (In progress) Improve the way client-side metadata is structured, make sure it matches server metadata shape, and try to avoid metadata duplication.
    
-   Update KSES to accept our link color styles.
    
-   Expose a structure in the core that wp-cli can use to know what theme.json strings are translatable.
    
-   Look into the alignment issues [https://github.com/WordPress/gutenberg/issues/26633](https://github.com/WordPress/gutenberg/issues/26633)
    
-   Discuss mobile support [#24165](https://github.com/WordPress/gutenberg/issues/24165) and have some double-checking if the current theme.json functionality can work well on mobile.
    

  
  
  

## What we can do next?

Tasks we can work on, it would be nice to have but are not a must and are lower priority may or may not be part of v1.

-   Create a JSON schema specification for theme.json. Developers can use the schema to validate if their theme.json files are correct and easily locate errors. Improving the developer experience of using theme.json. We can use the scheme in the core to remove invalid paths etc.
    

  

## Things for the future

Not part of a first version. Things we may consider in the future.

-   Nested styles for semantic contexts. For example, apply a specific base font size for things inside a sidebar.
    
-   Allow CSS properties that we don't manage at the moment. E.g: apply a specific css property
    
-   Responsive styles/media queries (styles depending on window size).
    

## What decisions do we need to do?

-   Are the APIs we offer themes enough to have control about which UI components are surfaced to the user in the global styles sidebar? OR should we expand our API’s?
-   If a block does not support something like line-height, but the theme specifies a line-height for that block. Should we output a line-height style anyway?
-   What are the controls that are available per block? Which blocks by default have font size available? Which have text color? Which have line-height?

  

## What’s next on the UI/UX side?

The first beta of WordPress 5.7 goes out in January, so we need to have a v1 for the UI ready sometime before the first beta.

-   Typography tools are going to be g2 based (or at least some parts like font size control). So we need to merge (at least parts of) g2 in order to have the typography tools ready.
    
-   Extensibility mechanism for 3rd party
    
-   What is the plan for the other parts of UI? What are the things we should improve on the UI of these other parts to make them ready for the first version?
    

	-   Navigation between all the different blocks that we can apply global styles to and between the global context.
    
	-   Color options (text, background, and link).
    
	-   Color palette editor.

**Comment by youknowriad** [comment]
Thanks for the summary @jorgefilipecosta here's the questions/feedback I have:

> (In progress) Improve the way client-side metadata is structured, make sure it matches server metadata shape, and try to avoid metadata duplication.

Shooting for consistency between client and server consistency is not always good thing.  It's better to think of client packages as their own unit in isolation.  Sometimes the config that the server has does make sense in the context of a package, sometimes it doesn't. Semantics > consistency. Just something to keep in mind while designing the APIs of the client.

> Look into the alignment issues #26633

this is also something I've been thinking of, It's not clear whether I'll have the time to look deeper into it but keep me on the loop. My idea is to first make InnerBlocks provide an API to define the alignments and output the CSS for it without relying on themes to provide that CSS. Once we have that, adding support for this into theme.json is easy.

> Responsive styles/media queries (styles depending on window size).

Same here, I wonder  if we should start at the block level IMO and see what makes  sense as a "style" attribute, and what make  sense as UI and expand to theme.json

> Are the APIs we offer themes enough to have control about which UI components are surfaced to the user in the global styles sidebar? OR should we expand our API’s? Should our API's allow to specify that the user can change a given style property like line-height at the global styles level, but not at the block level (block inspector).

I  still feel this is the missing piece in terms of features. 

  - How do I enable colors everywhere but only allow changing the main one in the Global styles Ui
  - How do I disable line-height choices in all blocks but allow it at the global level.

I see the current "global" config as a way to define a config for all blocks but not as a way to define what UI is available in the global styles panel.

**Comment by aristath** [comment]
Thank you @jorgefilipecosta, that comment offers a great overview and sums up pretty nicely where we're going. :heart: 
Are all items in separate issues already? The comment will be lost in all the noise, so should we create issues for any items on that list (or maybe just tag existing issues properly so they're easier to find)?

**Comment by francescamarano** [comment]
I second @aristath comment. It would also be great to see them in the FSE project so it's easy to track things visually. And make liberal use of tags, so people can work on priorities. Thanks!

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad,
Regarding alignment issues and responsive styles, I agree with you. These things should be solved at the block level without touching global style mechanics, at least initially. I moved them to a new section, called "Related tasks but that initially will not use theme.json/global styles mechanisms.".

> I see the current "global" config as a way to define a config for all blocks but not as a way to define what UI is available in the global styles panel.

Currently, the global config is a way to define what UI is available at the global styles panel. The other blocks inherit it if they don't have a specific config for them.

Basically, before global styles, one could not have control if a feature is enabled or disabled in a specific block. Our API changed that. With our new API, one can also specify if things in a particular block (or even in blocks with specific attributes) are available or not and overwrite what was defined globally. Our API just added block-level controls, but we don't have things like selectors that select multiple blocks, etc., ways to disable inheritance, etc. Themes always need to select each block individually.


> How do I enable colors everywhere but only allow changing the main one in the Global styles UI

We don't the concept of "main color". The user can edit, text color, background color, and link color, at the global level and at the level of blocks that support these colors. One can't enable text color but disable background color. All the colors are either enabled or disabled.
Besides that, the user can change the color presets, and if there is a color like "primary" used in multiple places, the user changing the preset will change the color in multiple places. It is not possible to say only this preset is editable. Do you think we should change some of these in the first version?


> How do I disable line-height choices in all blocks but allow it at the global level.

The current API does not offer a way to select all the blocks. Achieving this would require to enable line-height at the global level, and then explicitly disable it for all the block contexts. I guess it could be done using a PHP filter that loops through all the selectors and injects the setting to disable line-height keeping it enabled for the global context, which is not ideal.
I see two options to achieve:
- Have a selector similar to "*" that selects all the blocks? We enable line-height at the global level but disable it for all the blocks because all match the "*" selector.
- Have something like non inherited settings so we could use this to change line-height in global to true, making sure all the blocks below will not inherit the setting and have line-height enabled.

None of these options seem good. Do you have a different idea in mind?



There is a tradeoff between supporting more use cases and making theme.json structure and the engine implementing these rules more complex. I think in a declarative approach, there will always be things that will not be possible to achieve and we should draw a line. I guess we need to decide between what we allow and what we don't allow, and make any required engine updates as soon as possible.

Do you think there is a simple theme.json change we can make that would allow the use cases you referred? Are there other use cases besides the ones you already referred that you think should be possible?


Thank you a lot for looking into this and sharing your insights/questions!

**Comment by jorgefilipecosta** [comment]
Hi @aristath, @francescamarano,
Thank you for your comments! To make things easier to navigate, I updated the comment linking to the issues/PR's and also added more prioritary issues to the Full Site Editing board.

**Comment by youknowriad** [comment]
> Currently, the global config is a way to define what UI is available at the global styles panel. The other blocks inherit it if they don't have a specific config for them.

And I'm saying it shouldn't be the case, we should have two distinct things (both in theme.json): 

 - A way to disable and enable things for all blocks (without explicitly listing the blocks)
 - And a way to enable/disable things on the global panel UI

(maybe a "global"l key and a "UI" or "*" and "global" or something like that)

These things should be different. Once they are, the use-cases I shared will be solved. 

When I was referring to "main" color, for me this was just a custom CSS variable editable on the global styles UI and that you can use as a value of of the palette colors or not, you can also just use it to style some blocks in theme.json directly. (provide default values)

**Comment by cliffordp** [comment]
Just chiming in to say https://wordpress.org/plugins/kt-tinymce-color-grid/ is an excellent plugin that works for Classic, Block, and Customizer. It adds CSS variables with the name you choose for each color (e.g. brand-main, brand-secondary, brand-accent, sheepish-white, creamy-gray, deep taupe), and it has import/export functionality.

I hope it can inspire.

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad, thank you for clarifying your thoughts.

> And I'm saying it shouldn't be the case, we should have two distinct things (both in theme.json):

> A way to disable and enable things for all blocks (without explicitly listing the blocks)
> And a way to enable/disable things on the global panel UI
(maybe a "global"l key and a "UI" or "*" and "global" or something like that)



I guess we can accomplish these requirements in two possible ways:


- Two selectors a "*" selector that is inherited by every context (including the global one). Equivalent to our current "global". A global context that just affects the global panel and is not inherited. There is no explicit inheritance of styles so styles would still be defined on "global". "*" Selector would just support settings which is the only place where we are inhering things right now.

- Keeping everything as it is, but add a new special settings only context called "global-panel". This panel just affects global context settings without any inheritance. On the global context, if this context contains settings it overwrites what was defined in global.


Do you think these solutions are acceptable? Do you have any preferences? Also cc: @nosolosw, in case you have some thoughts.


> When I was referring to "main" color, for me this was just a custom CSS variable editable on the global styles UI and that you can use as a value of of the palette colors or not, you can also just use it to style some blocks in theme.json directly. (provide default values).


Well, one can edit presets, they are variables. One can have a preset called the main color. Editing that preset would be equivalent to edit a variable used all around the site.
But besides the color presets, and, the link color (which is also a CSS variable), there are no plans as part of Global Styles V1 to have a mechanism that allows someone to expose a UI in the global styles panel that allows controlling the value of a CSS variable. Do you think need to have for V1 something that allows controlling custom CSS variables (besides color presets and link color)?

**Comment by jorgefilipecosta** [comment]
> Just chiming in to say [wordpress.org/plugins/kt-tinymce-color-grid](https://wordpress.org/plugins/kt-tinymce-color-grid/) is an excellent plugin that works for Classic, Block, and Customizer. It adds CSS variables with the name you choose for each color (e.g. brand-main, brand-secondary, brand-accent, sheepish-white, creamy-gray, deep taupe), and it has import/export functionality.

Hi @cliffordp,

Thank you for bringing this plugin to our attention. It has considerable usage, which means users really want the power of changing color palettes and things like that, so it reinforces that we are working on something relevant.
It is also a good test case to make sure what we are doing is back-compatible and does not breaks plugins. E.g: that plugin should still work and the colors users apply with it should still be applied on global styles (unless the user then changes the colors on global styles).

**Comment by youknowriad** [comment]
> Do you think these solutions are acceptable? Do you have any preferences? Also cc: @nosolosw, in case you have some thoughts.

Yes, I think this is good. I'd love more thoughts on my intuition here though, especially from theme authors, let's not rush implementation.

>  Do you think need to have for V1 something that allows controlling custom CSS variables (besides color presets and link color)?

Same here, I do think it's a valuable addition especially since you can use this value in different places in theme.json. But I agree it may not be important for V1.

**Comment by nosolosw** [comment]
@youknowriad @jorgefilipecosta I've taken your comments here, created an issue to discuss this separately, and added my own thoughts as well. See https://github.com/WordPress/gutenberg/issues/27295

**Comment by jorgefilipecosta** [comment]
It is time for another update on Global Styles :)

## Updates on the must-have items

We got some traction on the i18n of theme.json:
https://github.com/WordPress/gutenberg/pull/27380
We hope to merge this PR soon. And that should conclude the i18n work of theme.json on the Gutenberg side. It still needs work on wp-cli side to extract the strings. Now we expose a structure that can be consumed.

We also have some updates on the client-side metadata:
https://github.com/WordPress/gutenberg/pull/27453 is ready for review.
https://github.com/WordPress/gutenberg/pull/27449 was merged.
When we merge the remaining PR, that part of the work for v1 is done.


Regarding child theme support, that work was put on on pause as @nosolosw referred:
> I've seen there's rapid movement in the FSE/templates area and I believe when that's more settled I can resume this work.

## Design tools visibility discussion

Regarding the question we had on the last update:
> Are the APIs we offer themes enough to have control about which UI components are surfaced to the user in the global styles sidebar? OR should we expand our API’s? Should our API's allow to specify that the user can change a given style property like line-height at the global styles level, but not at the block level (block inspector).

There were some discussions on this issue, and an issue for it was created https://github.com/WordPress/gutenberg/issues/27295. Solving this challenge and updating the theme.json shape for this is the biggest priority we have right now. If you have insights for this discussion, please share them on the issue.

## UI/UX
The big question we had on the last update was the UI/UX, and on that front, there are important updates.
The first version should implement the prototype that is available at https://g2-components.xyz/iframe.html?id=examples-wip-globalstylessidebar--default&viewMode=story#.
It will involve enhancing many of the current components we have to offer that UI.
The issue to track this UI update is available at https://github.com/WordPress/gutenberg/issues/27473. And the part related to the specific components is available at https://github.com/WordPress/gutenberg/issues/27331.
After this update, there are several small enhancements we can make to make the user experience better:
- https://github.com/WordPress/gutenberg/issues/27393
- https://github.com/WordPress/gutenberg/issues/27474
- https://github.com/WordPress/gutenberg/issues/27475


Given that we already have considerable progress on the infrastructure part, the UI is considered a high priority. We will give a bigger focus to the UI work. The only two things considered higher priority are the i18n work and the issue https://github.com/WordPress/gutenberg/issues/27295 already referred.

**Comment by nosolosw** [comment]
:wave: Here's the current status for this task.

Latest changes:

- Gutenberg 9.9 will come with a new format for the theme.json, see [docs](https://developer.wordpress.org/block-editor/developers/themes/theme-json/). https://github.com/WordPress/gutenberg/pull/28110 https://github.com/WordPress/gutenberg/pull/28533
- The TT1-blocks theme has been updated to the new format, although not published in the WordPress repo yet. https://github.com/WordPress/theme-experiments/pull/182/ https://github.com/WordPress/theme-experiments/pull/175
- Presets:
  - Set up was added to translate them when declared via theme.json. It won't be necessary to declare them in functions.php as well when this lands in WordPress core. It still is necessary while we iterate in the plugin. https://github.com/WordPress/gutenberg/pull/27380
  - Font style, weight, decoration, and transforms are no longer presets. https://github.com/WordPress/gutenberg/pull/27555
- Support for new properties at the infrastructure level: border-radius, border-color, border-width, border-style. https://github.com/WordPress/gutenberg/pull/28049 https://github.com/WordPress/gutenberg/pull/27791 https://github.com/WordPress/gutenberg/pull/25791
- UI controls: added padding to the global sidebar. https://github.com/WordPress/gutenberg/pull/27154
- Custom spacing has been ported to WordPress core and will be part of 5.7 https://github.com/WordPress/wordpress-develop/pull/951
- Continuous documentation updates, fixes, and general polishing.
- Many blocks were updated to expose more UI controls for style properties.

What's next?

- Immediate: list the new UI controls for styles as a "dev note" for 5.7, so themes are aware of the new things users can tweak https://github.com/WordPress/gutenberg/issues/28539
- Immediate: finish some loose ends https://github.com/WordPress/gutenberg/issues/27506 https://github.com/WordPress/gutenberg/issues/28163
- We've been working towards shipping parts of theme.json to 5.7 https://github.com/WordPress/gutenberg/issues/27506 While there was a lot of progress, it proved premature given two outstanding issues that are unresolved in the format: how to express different templates/nested contexts & support the new direction for sidebars (controls that are shown/hidden) https://github.com/WordPress/gutenberg/issues/27331 Addressing these issues would be our next focus.

Expect some issue triage in the coming days, as well as updates to this issue description.

**Comment by nosolosw** [comment]
For the people following this task's updates: the issue description has been updated and there's now a third iteration that includes an initial set of tasks to be developed for April's MVP. All the rest has been moved to the backlog or is done.

**Comment by nosolosw** [comment]
Update on this: the core feature we wanted for block supports (see https://github.com/WordPress/gutenberg/issues/28913) to be able to absorb more use cases is now working on the buttons block (color hook). There're some loose ends I plan to keep working on in the coming days. There are also more hooks and blocks to support and I listed them as individual tasks on 28913, in case anyone finds time to help with them).

**Comment by nosolosw** [comment]
Update for this week's core-editor meeting:

- Block supports: landed the first implementation that doesn't serialize the block attributes. The overview task lists other properties we should migrate and potential blocks https://github.com/WordPress/gutenberg/issues/28913 **This is a good place to start looking at for anyone interested in helping with global styles.**

- Internationalization. There's going to be some work in the translation areas for theme.json & block supports in the next weeks (@gziolo). See https://github.com/WordPress/gutenberg/pull/28783 and https://github.com/WordPress/gutenberg/pull/29155

- theme.json. I'm going to resume work on some of the things we wanted to do for theme.json in the next few days https://github.com/WordPress/gutenberg/issues/28163

**Comment by nosolosw** [comment]
There're some changes coming to the theme.json format as per https://github.com/WordPress/gutenberg/issues/29891 I'll update this issue's description and the affected tasks in the next few days.

**Comment by oandregal** [comment]
Update for this week core-editor chat:

- The [last round of mockups for the global styles sidebar](https://github.com/WordPress/gutenberg/issues/27473) are live. It has provided a good ground to re-evaluate the theme.json format and there's going to be some formalization of the concept of "elements", see https://github.com/WordPress/gutenberg/issues/29891 for details.
- There's ongoing work for translations https://github.com/WordPress/gutenberg/pull/29828 and iterate to offer higher-level APIs in the server https://github.com/WordPress/gutenberg/pull/29667 https://github.com/WordPress/gutenberg/pull/29905
- We landed some fixes for sub-properties (padding) as well https://github.com/WordPress/gutenberg/pull/29712

This issue description has been updated to focus on the infrastructure parts of global styles, that we aim to ship in 5.8. The UI tasks have been moved to the backlog for clarity but will be re-evaluated at a later point.

**Comment by oandregal** [comment]
This week in global styles:

- Merged a few important things for theme.json: add a [layout](https://github.com/WordPress/gutenberg/pull/29335) config in theme.json, [allow themes to use any style via theme.json whether or not the block supports it](https://github.com/WordPress/gutenberg/pull/29941), translation for [custom templates](https://github.com/WordPress/gutenberg/pull/29828).
- Continue iterating on the block supports mechanism: [allow skipping the serialization of border](https://github.com/WordPress/gutenberg/pull/30035).
- Some bugfixes: [30088](https://github.com/WordPress/gutenberg/pull/30088).

Ongoing/Next tasks necessary for MVP:

- theme.json: [changes to the format](https://github.com/WordPress/gutenberg/issues/29891)
- block supports: [make the button block use the new mechanism](https://github.com/WordPress/gutenberg/pull/30035#issuecomment-805184310)
- internationalization: labels for UI controls [29155](https://github.com/WordPress/gutenberg/pull/29155) [30075](https://github.com/WordPress/gutenberg/pull/30075)
- extensibility: [discussion about the existing filters and how we should enable hooks](https://github.com/WordPress/gutenberg/issues/30082)

And more:

- [REST Endpoint for mobile](https://github.com/WordPress/gutenberg/pull/29969)
- [Expand border support to color, radius, and width](https://github.com/WordPress/gutenberg/pull/30124/files)
- [Add support for child themes in global styles (not FSE)](https://github.com/WordPress/gutenberg/pull/30147)
- [Global Styles label](https://github.com/WordPress/gutenberg/labels/Global%20Styles).

**Comment by oandregal** [comment]
Just added https://github.com/WordPress/gutenberg/issues/30191 to the list as well (under the theme.json group).

**Comment by oandregal** [comment]
This week's status report: ongoing work for the major tasks necessary for the MVP.

- Continue to prepare the changes for theme.json https://github.com/WordPress/gutenberg/issues/29891
- Conversations about the settings for the different editors and mobile
  - https://github.com/WordPress/gutenberg/issues/30082
  - https://github.com/WordPress/gutenberg/pull/29969
  - https://github.com/WordPress/wordpress-develop/pull/1118
  - https://github.com/WordPress/gutenberg/pull/30245

**Comment by oandregal** [comment]
This week's status report:

**Shipped**

- Block Supports: added a `gutenberg_block_has_support` helper function and support for padding in dynamic blocks [30322](https://github.com/WordPress/gutenberg/pull/30332), border for search block https://github.com/WordPress/gutenberg/pull/30227

- Fixes: title in the save panel of the site editor [30521](https://github.com/WordPress/gutenberg/pull/30521), fix for themes without theme.json support but with experimental link color support [30452](https://github.com/WordPress/gutenberg/pull/30452)

**Ongoing**

- Block supports:
	- border for button block https://github.com/WordPress/gutenberg/pull/30194, 
	- add border color/style/width properties https://github.com/WordPress/gutenberg/pull/30124
	- spacing https://github.com/WordPress/gutenberg/pull/30333 https://github.com/WordPress/gutenberg/pull/30371

- Theme.json
  - Update format https://github.com/WordPress/gutenberg/issues/29891 There's now a draft PR at https://github.com/WordPress/gutenberg/pull/30541
  - Conversation about whether and how to expand the use of CSS Custom Properties https://github.com/WordPress/gutenberg/pull/29714

**Comment by oandregal** [comment]
This week in global styles:

**Landed**

- Parts of theme.json with elements https://github.com/WordPress/gutenberg/issues/29891
	- preparations to make changes easier https://github.com/WordPress/gutenberg/pull/30801 https://github.com/WordPress/gutenberg/pull/30610 
	- i18n process: make it not depend on a particular path https://github.com/WordPress/gutenberg/pull/30604
	- add mechanism to migrate from a v0 schema to the latest https://github.com/WordPress/gutenberg/pull/30600
	- clean-ups https://github.com/WordPress/gutenberg/pull/30605 https://github.com/WordPress/gutenberg/pull/30612

- block supports
	- border radius for the button block https://github.com/WordPress/gutenberg/pull/30194
	- font size & font family for the button block https://github.com/WordPress/gutenberg/pull/30394
	- colors for the verse block https://github.com/WordPress/gutenberg/pull/27736
	- border radius for the search block https://github.com/WordPress/gutenberg/pull/30227

**Ongoing**

- theme.json with elements https://github.com/WordPress/gutenberg/issues/29891
	- extract sanitization method https://github.com/WordPress/gutenberg/pull/30809
	- changes to shape https://github.com/WordPress/gutenberg/pull/30541

- bugfixes https://github.com/WordPress/gutenberg/pull/30830/

- block supports: add new style properties: border-color, border-styles, border-width https://github.com/WordPress/gutenberg/pull/30124

**Comment by oandregal** [comment]
This week in global styles:

**Landed**

theme.json with elements https://github.com/WordPress/gutenberg/issues/29891

- extract sanitization method https://github.com/WordPress/gutenberg/pull/30809

Block Supports & theme.json properties:

- Added block support & theme.json support for: border-color, border-style, and border-width. https://github.com/WordPress/gutenberg/pull/30124

Bugfixes, code quality

- Clean cached data when switching themes https://github.com/WordPress/gutenberg/pull/30830/
- Standarize block editor settings load https://github.com/WordPress/gutenberg/pull/30245
- Fix regression in color panel of global styles sidebar https://github.com/WordPress/gutenberg/pull/31015
- Add utility to retrieve classes and styles in color hook https://github.com/WordPress/gutenberg/pull/30869 https://github.com/WordPress/gutenberg/pull/30870
- Polish color labels https://github.com/WordPress/gutenberg/pull/30075

**Ongoing**

theme.json with elements https://github.com/WordPress/gutenberg/issues/29891

- changes to shape (working in all editors & front-end) https://github.com/WordPress/gutenberg/pull/30541

Block Supports:

- Allow skipping serialization for typography attributes https://github.com/WordPress/gutenberg/pull/30880
- Conversation about "overlay" colors https://github.com/WordPress/gutenberg/issues/29963#issuecomment-820494790
- Migrate more blocks using the skipSerialization mechanism as well as adding more properties https://github.com/WordPress/gutenberg/issues/28913

Bugfixes:

- kses filtering (approved) https://github.com/WordPress/gutenberg/pull/30888

**Priorities**

The priority continues to be shipping https://github.com/WordPress/gutenberg/issues/29891 as soon as possible to unblock related work.

https://github.com/WordPress/gutenberg/issues/28913 is a good issue to look at if you can lend a hand (help with reviews, migrate more blocks, more properties, etc).

**Comment by oandregal** [comment]
Weekly status report for [core-editor meeting](https://make.wordpress.org/core/2021/04/26/editor-chat-agenda-28-april-2021/).

**Shipped this week**

theme.json

- clean-ups https://github.com/WordPress/gutenberg/pull/31224

block.json / block supports

- duotone https://github.com/WordPress/gutenberg/pull/26752
- padding: fix to allow units other than px https://github.com/WordPress/gutenberg/pull/31057
- font-size: can be used in nodes other than the wrapper https://github.com/WordPress/gutenberg/pull/30879
- border-color: fix for dynamic blocks https://github.com/WordPress/gutenberg/pull/31020
- blocks
	- padding for site title https://github.com/WordPress/gutenberg/pull/31125
	- padding and letter casing for site tagline https://github.com/WordPress/gutenberg/pull/31042
	- link color for post date https://github.com/WordPress/gutenberg/pull/30791
	- table uses now background/text colors from the theme https://github.com/WordPress/gutenberg/pull/30791

Internationalization

- custom templates can be translated via theme.json https://github.com/WordPress/gutenberg/pull/29828

KSES

- fix to sanitizing user styles https://github.com/WordPress/gutenberg/pull/30888

Site editor:

- fixes for preset classes in the site editor https://github.com/WordPress/gutenberg/pull/31218 https://github.com/WordPress/gutenberg/pull/31217 https://github.com/WordPress/gutenberg/pull/31204
- fix for color selector in the UI color panel https://github.com/WordPress/gutenberg/pull/31015

**Ongoing**

theme.json

- Changes to format https://github.com/WordPress/gutenberg/issues/29891 are ready https://github.com/WordPress/gutenberg/pull/30541 we're now testing and polishing. It also includes versioning (aka migrate the older formats to the new one).
- Improve API to take data from the theme directly https://github.com/WordPress/gutenberg/pull/31267

block.json / block supports

- typography (line-height, font-weight, etc) can be used in nodes other than the wrapper https://github.com/WordPress/gutenberg/pull/30880
- border and color (background, text) to the pull quote block https://github.com/WordPress/gutenberg/pull/30951

New style properties

- Add support for letter-spacing https://github.com/WordPress/gutenberg/pull/31118

Misc

- Add a new endpoint that exposes block editor settings through the REST API (by mobile to access global styles data) https://github.com/WordPress/gutenberg/pull/29969
- standardize capabilities of wp_template, wp_template_part and wp_global_styles https://github.com/WordPress/gutenberg/pull/30893


**Priorities**

The priority continues to be shipping https://github.com/WordPress/gutenberg/issues/29891 as soon as possible to unblock related work.

https://github.com/WordPress/gutenberg/issues/28913 is a good issue to look at if you can lend a hand (help with reviews, migrate more blocks, more properties, etc).

**Comment by oandregal** [comment]
Status report for [core-editor meeting](https://make.wordpress.org/core/2021/05/03/core-editor-meeting-agenda-5-may-2021/).

**Shipped this week:**

block.json / block supports:

- Improve the typing performance by only rendering hook panel of selected blocks https://github.com/WordPress/gutenberg/pull/31381
- Update padding to allow configurable sides https://github.com/WordPress/gutenberg/pull/30607
- Add utils to get border classes and inline styles https://github.com/WordPress/gutenberg/pull/31264

theme.json

- Change the format of theme.json and introduce elements as well as versioning https://github.com/WordPress/gutenberg/pull/30541
- The top-level styles & settings use the `body` selector instead of `:root` https://github.com/WordPress/gutenberg/pull/31302

Extensibility:

- Global Styles settings to use the `block_editor_settings_all` filter https://github.com/WordPress/gutenberg/pull/31027

Site editor:

- Improve APIs to allow getting theme styles https://github.com/WordPress/gutenberg/pull/31267

**Ongoing:**

block.json / block supports:

- typography: allow blocks to use these properties in inner markup https://github.com/WordPress/gutenberg/pull/30880
- border: allow non-pixel units https://github.com/WordPress/gutenberg/pull/31483
- margin: allow configurable sides https://github.com/WordPress/gutenberg/pull/30608
- add letter-spacing https://github.com/WordPress/gutenberg/pull/31118
- migrate more blocks to the hooks system:
	- border for cover https://github.com/WordPress/gutenberg/pull/31370 and image https://github.com/WordPress/gutenberg/pull/31366
	- margin for separator https://github.com/WordPress/gutenberg/pull/30609\
	- border & color for pullquote https://github.com/WordPress/gutenberg/pull/30951

theme.json

- Use theme taxonomy instead of post name to associate user data (stored as a CPT) with the theme https://github.com/WordPress/gutenberg/pull/31436

**Priorities**

After having landed https://github.com/WordPress/gutenberg/issues/29891 via https://github.com/WordPress/gutenberg/pull/30541 the priority is to tie up the loose ends before preparing the core patch for 5.8.

Worth highlighting is that there're two new issues to switching away from CSS Custom Properties for link https://github.com/WordPress/gutenberg/issues/31488 and font-family https://github.com/WordPress/gutenberg/issues/31489 (see rationale in each issue).

**Comment by oandregal** [comment]
This week report for [core-editor](https://make.wordpress.org/core/2021/05/10/core-editor-agenda-12-may-2020/).

**Shipped this week**

block supports

- margin: configurable sides https://github.com/WordPress/gutenberg/pull/30608
- spacing: hide UI for disabled properties https://github.com/WordPress/gutenberg/pull/31726
- rename and stabilize `useEditorFeature` to `useSetting` https://github.com/WordPress/gutenberg/pull/31587
- layout: add reset button https://github.com/WordPress/gutenberg/pull/30828

theme.json

- update shape to include elements https://github.com/WordPress/gutenberg/pull/30541
- fix settings retrieval for REST https://github.com/WordPress/gutenberg/pull/31534
- use wp_theme taxonomy for CPT https://github.com/WordPress/gutenberg/pull/31436
- rename `experimental-theme.json` to `theme.json` https://github.com/WordPress/gutenberg/pull/29981
- code quality https://github.com/WordPress/gutenberg/pull/31552 https://github.com/WordPress/gutenberg/pull/31666 https://github.com/WordPress/gutenberg/pull/31596 https://github.com/WordPress/gutenberg/pull/31573 

site editor:

- cover against blocks being unregister https://github.com/WordPress/gutenberg/pull/31588
- create metadata for all blokcs https://github.com/WordPress/gutenberg/pull/31590
- remove specificity for link color https://github.com/WordPress/gutenberg/pull/31497

**Ongoing**

block supports:

- typography: allow blocks to use these properties in inner markup https://github.com/WordPress/gutenberg/pull/30880
- border: allow non-pixel units https://github.com/WordPress/gutenberg/pull/31483
- add letter-spacing https://github.com/WordPress/gutenberg/pull/31118
- migrate more blocks to the hooks system:
	- border for cover https://github.com/WordPress/gutenberg/pull/31370 and image https://github.com/WordPress/gutenberg/pull/31366
	- margin for separator https://github.com/WordPress/gutenberg/pull/30609\
	- border & color for pullquote https://github.com/WordPress/gutenberg/pull/30951

theme.json

- Show both core & theme colors https://github.com/WordPress/gutenberg/pull/31669
- Link color not to use CSS Custom Property https://github.com/WordPress/gutenberg/pull/31524

**Priorities**

Start the merge process for theme.json APIs.

**Comment by oandregal** [comment]
Status report ([core-editor](https://make.wordpress.org/core/2021/05/17/editor-chat-agenda-19-may-2021/)):

This week's focus was on getting things ready for Gutenberg 10.7 (landed most of the things we needed, there are a couple in flight) and start the merge process in core (got an initial PR for this).

**Shipped this week**

block supports:

- margin: fix for custom units https://github.com/WordPress/gutenberg/pull/31776
- margin: add server-side support https://github.com/WordPress/gutenberg/pull/31808
- duotone: fix padding crash https://github.com/WordPress/gutenberg/pull/31780
- font-family: fix typo that affects dynamic blocks https://github.com/WordPress/gutenberg/pull/31974
- layout: fix for complex values https://github.com/WordPress/gutenberg/pull/31740
- typography: allow blocks to use these properties in inner markup https://github.com/WordPress/gutenberg/pull/30880
- link color: do not set default https://github.com/WordPress/gutenberg/pull/31755
- link color: fix selector https://github.com/WordPress/gutenberg/pull/31820

settings:

- REST Request for mobile: more data https://github.com/WordPress/gutenberg/pull/31819 https://github.com/WordPress/gutenberg/pull/31762

**Ongoing**

block supports:

- typography: allow blocks to use these properties in inner markup https://github.com/WordPress/gutenberg/pull/30880
- border: allow non-pixel units https://github.com/WordPress/gutenberg/pull/31483
- add letter-spacing https://github.com/WordPress/gutenberg/pull/31118
- migrate more blocks to the hooks system:
	- border for cover https://github.com/WordPress/gutenberg/pull/31370 and image https://github.com/WordPress/gutenberg/pull/31366
	- margin for separator https://github.com/WordPress/gutenberg/pull/30609\
	- border & color for pullquote https://github.com/WordPress/gutenberg/pull/30951

theme.json

- Show both core & theme colors https://github.com/WordPress/gutenberg/pull/31669
- Link color not to use CSS Custom Property https://github.com/WordPress/gutenberg/pull/31524

core merge

- Add settings for all editors https://github.com/WordPress/wordpress-develop/pull/1262

**Priorities**

Merge process for theme.json APIS in core.

**Comment by oandregal** [comment]
Status Report ([core-editor meeting](https://make.wordpress.org/core/2021/05/26/editor-chat-agenda-26-may-2021/#comment-41358)).

Global Styles and Global Settings have been ported to WordPress core and will be part of the WordPress 5.8 release. The focus now is on polishing and fixing bugs as we find them. The issue description & tasks have been triaged and there're two new sections: "bugfixes" and "docs" with tasks to be worked on in the upcoming weeks.

Now that the 1.0 version of this feature has been shipped and we need to revisit focuses, I thought it'd be a good time to **pause these reports** while we figure out the next steps. Anything relevant should still be posted here but on an on-demand basis not weekly.

**Supported settings & styles that landed core**

There's a task above to update this in the docs but wanted to share what we have now.

- The following settings & styles are still experimental and only work with the plugin: border, font-family, font-style, font-weight, text-decoration, text-transform. The rest are part of WordPress core now, so they work without the plugin as well.
- Layout is enabled for themes with `theme.json` support.
- The default theme.json settings we landed in core work like this:
  - If the settings had a corresponding [theme support](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-support/), it works the same as before. Settings that are enabled by default are `color.custom`, `color.customGradient`, `typography.customFontSize`, `typography.dropCap`. Settings that are disabled by default are `spacing.customMargin`, `spacing.customPadding`, `typography.lineHeight`. Units can be filtered.
  -  Link color is stable, landed into 5.8, and it's disabled by default. Need to make a decision on what to do with the existing `add_theme_support('experimental-link-color')`. It seems that we can keep it in the plugin and remove it when the minimum WordPress version is 5.8.

**Shipped this week**

Port to core:

- core: [settings](https://github.com/WordPress/wordpress-develop/pull/1287), [styles](https://github.com/WordPress/wordpress-develop/pull/1287)
- gutenberg: [adapt core](https://github.com/WordPress/gutenberg/pull/32059/)

Block Supports:

- link color: refactored to use the elements mechanism https://github.com/WordPress/gutenberg/pull/31524
- typography: allow blocks to use these properties in inner markup https://github.com/WordPress/gutenberg/pull/30880
- spacing (margin, padding): allow blocks to use these properties in inner markup https://github.com/WordPress/gutenberg/pull/31973
- font-family: fix for dynamic blocks https://github.com/WordPress/gutenberg/pull/31974
- duotone: add an option to disable it https://github.com/WordPress/gutenberg/pull/32002
- units: make UI controls use the ones defined via theme.json https://github.com/WordPress/gutenberg/pull/31822

theme.json

- Make core colors (classes and CSS Custom Properties) always available https://github.com/WordPress/gutenberg/pull/31669 This landed in core as well, but we've found some issues such as https://github.com/WordPress/gutenberg/pull/32225 so we're evaluating what to do.

Misc:

- Docs: make examples copy-pasteable https://github.com/WordPress/gutenberg/pull/32040
- Fix for REST endpoint used by mobile https://github.com/WordPress/gutenberg/pull/31844

**Comment by oandregal** [comment]
We're shipping a first version of the GS interface in 5.9, so it looks like this issue could be closed. Both for having a sense of progress but also because the activity has already moved to other smaller and more focused issues (designer tools, site editor improvements, framework changes, etc).

**Comment by oandregal** [comment]
I've created https://github.com/WordPress/gutenberg/issues/37550 to track some gaps we still have from the original vision of styles at the framework level.

**Comment by mtias** [comment]
Really wonderful work and progress here!
